channeldb/channel.go

```diff
@@ -1364,6 +1371,51 @@ func (c *OpenChannel) SecondCommitmentPoint() (*btcec.PublicKey, error) {
 	return input.ComputeCommitmentPoint(revocation[:]), nil
 }
 
+// MarkShutdownSent is used to mark that we've sent a Shutdown message for this
+// channel. This is so we can retransmit Shutdown.
+func (c *OpenChannel) MarkShutdownSent() error {
+	c.Lock()
+	defer c.Unlock()
+
+	if err := kvdb.Update(c.Db.backend, func(tx kvdb.RwTx) error {
+		chanBucket, err := fetchChanBucketRw(
+			tx, c.IdentityPub, &c.FundingOutpoint, c.ChainHash,
+		)
+		if err != nil {
+			return err
+		}
+
+		channel, err := fetchOpenChannel(
+			chanBucket, &c.FundingOutpoint,
+		)
+		if err != nil {
+			return err
+		}
+
+		channel.sentShutdown = true
+
+		return putOpenChannel(chanBucket, channel)
+	}, func() {}); err != nil {
+		return err
+	}
+
+	c.sentShutdown = true
+
+	return nil
+}
+
+// HasSentShutdown returns whether or not we've ever sent Shutdown for this
+// channel. For nodes upgrading to 0.16.0, this will initially be false even if
+// a Shutdown has been sent. This doesn't matter since this function is only
+// used when restarting a ChannelLink and pre-0.16.0 nodes only sent Shutdown
+// after the link was permanently stopped.
```

Comment:
> after the link was permanently stopped.

"permanently stopped" meaning the node's link won't be restarted?

Reply:
yes, master will mark the channel as coop closed in the database so the link can't be reloaded once coop close is initiated